kolibri/auth/test/test_permissions.py

```diff
@@ -1,454 +1,666 @@
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.remove_facility_admin  ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_REMOVE_FACILITY_ADMIN = 'auth.remove_facility_admin'
-
-#     def test_remove_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
+"""
+Tests of the permissions on specific models in the auth app. For tests of the permissions system itself, see test_permission_classes.py
+"""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from django.test import TestCase
+
+from .helpers import create_dummy_facility_data
+
+from ..constants import role_kinds
+from ..filters import HierarchyRelationsFilter
+from ..models import DeviceOwner, Facility, Classroom, LearnerGroup, Role, Membership, FacilityUser
+
+class ImproperUsageIsProperlyHandledTestCase(TestCase):
+    """
+    Tests that error cases and misuse of the interface are properly caught.
+    """
+
+    def setUp(self):
+        self.data1 = create_dummy_facility_data()
+        self.data2 = create_dummy_facility_data()
+        self.device_owner = DeviceOwner.objects.create(username="boss")
+
+    def test_that_checking_creation_perms_on_invalid_model_returns_false(self):
+        # cannot create a LearnerGroup with invalid attribute name
+        self.assertFalse(self.data1["facility_admin"].can_create(LearnerGroup, {"bad_attr_name": 77, "parent": self.data1["facility"]}))
+        # cannot create a LearnerGroup with missing attribute value ("name")
+        self.assertFalse(self.data1["facility_admin"].can_create(LearnerGroup, {"parent": self.data1["facility"]}))
+
+    def test_that_getting_roles_for_noncollection_fails(self):
+        with self.assertRaises(ValueError):
+            self.data1["facility_admin"].get_roles_for(object())
+        with self.assertRaises(ValueError):
+            self.data1["facility_admin"].has_role_for([role_kinds.ADMIN], object())
+
+    def test_that_getting_roles_for_deviceowner_returns_false(self):
+        self.assertFalse(self.data1["facility_admin"].has_role_for_user([role_kinds.ADMIN], self.device_owner))
+
+    def test_that_getting_roles_for_user_in_other_facility_returns_false(self):
+        self.assertFalse(self.data1["facility_admin"].has_role_for_user([role_kinds.ADMIN], self.data2["learners_one_group"][0][0]))
+
+    def test_that_invalid_references_to_hierarchyrelationsfilter_throw_errors(self):
+        with self.assertRaises(Exception):
```

Comment:
Going on my earlier comment about raising plain `Exceptions`, IMO the result is that now you expect to *catch* them somewhere. Since you're now catching every error, even those you may not have intended, it could suppress a legitimate mistake later on.